{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e745b02d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import biotda.mapper_pipline as mapper_pipline\n",
    "import biotda.preprocessing as preprocessing\n",
    "import numpy as np\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import json\n",
    "import biotda.main as main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561ea5dd",
   "metadata": {},
   "source": [
    "## Single-nucleus chromatin accessibility and transcriptomic characterization of Alzheimer’s Disease (GSE174367)\n",
    "\n",
    "Experiment type:\tExpression profiling by high throughput sequencing. Genome binding/occupancy profiling by high throughput sequencing\n",
    "\n",
    "The gene-regulatory landscape of the brain is highly dynamic in health and disease, coordinating a menagerie of biological processes across¬ distinct cell-types. Here, we present a multi-omic single-nucleus study of 191,890 nuclei in late-stage Alzheimer’s Disease (AD), accessible through our intuitive web portal, profiling chromatin accessibility and gene expression in the same biological samples and uncovering vast cellular heterogeneity. We identified cell-type specific, disease-associated candidate cis-regulatory elements and their candidate target genes, including an oligodendrocyte-associated regulatory module containing links to APOE and CLU. We describe cis-regulatory relationships in specific cell-types at AD risk loci defined by genome wide association studies (GWAS), demonstrating the utility of this multi-omic single-nucleus approach. Trajectory analysis of glial populations identified disease-relevant transcription factors, like SREBF1 and their regulatory targets. Further, we introduce scWGCNA, a co-expression network analysis strategy robust to the sparsity of single-cell data, to perform a systems-level meta-analysis of AD transcriptomics.\n",
    " \t\n",
    "Morabito S, Miyoshi E, Michael N, Shahin S et al. Single-nucleus chromatin accessibility and transcriptomic characterization of Alzheimer's disease. Nat Genet 2021 Aug;53(8):1143-1155. PMID: 34239132"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c11a4535",
   "metadata": {},
   "source": [
    "### Reading preprocessed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "94edb56b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000283597</th>\n",
       "      <th>ENSG00000283619</th>\n",
       "      <th>ENSG00000283623</th>\n",
       "      <th>ENSG00000283633</th>\n",
       "      <th>ENSG00000283635</th>\n",
       "      <th>ENSG00000283646</th>\n",
       "      <th>ENSG00000283652</th>\n",
       "      <th>ENSG00000283662</th>\n",
       "      <th>ENSG00000283674</th>\n",
       "      <th>ENSG00000283683</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sample-1</th>\n",
       "      <td>7.316251</td>\n",
       "      <td>4.274548</td>\n",
       "      <td>7.262058</td>\n",
       "      <td>6.663354</td>\n",
       "      <td>5.221161</td>\n",
       "      <td>6.813789</td>\n",
       "      <td>6.806842</td>\n",
       "      <td>9.458696</td>\n",
       "      <td>10.129412</td>\n",
       "      <td>7.288785</td>\n",
       "      <td>...</td>\n",
       "      <td>4.340793</td>\n",
       "      <td>5.641259</td>\n",
       "      <td>3.754220</td>\n",
       "      <td>7.674109</td>\n",
       "      <td>7.034334</td>\n",
       "      <td>5.624611</td>\n",
       "      <td>9.289292</td>\n",
       "      <td>6.550443</td>\n",
       "      <td>7.916829</td>\n",
       "      <td>4.786196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-10</th>\n",
       "      <td>6.808974</td>\n",
       "      <td>5.260135</td>\n",
       "      <td>7.115016</td>\n",
       "      <td>7.203029</td>\n",
       "      <td>4.165419</td>\n",
       "      <td>6.460773</td>\n",
       "      <td>6.812336</td>\n",
       "      <td>9.905395</td>\n",
       "      <td>10.617105</td>\n",
       "      <td>7.933976</td>\n",
       "      <td>...</td>\n",
       "      <td>3.712015</td>\n",
       "      <td>5.428915</td>\n",
       "      <td>4.434448</td>\n",
       "      <td>8.343289</td>\n",
       "      <td>7.447873</td>\n",
       "      <td>5.615707</td>\n",
       "      <td>8.571755</td>\n",
       "      <td>6.627103</td>\n",
       "      <td>7.709677</td>\n",
       "      <td>5.008819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-100</th>\n",
       "      <td>6.552535</td>\n",
       "      <td>4.804724</td>\n",
       "      <td>6.923650</td>\n",
       "      <td>6.245148</td>\n",
       "      <td>5.204919</td>\n",
       "      <td>6.162263</td>\n",
       "      <td>6.554665</td>\n",
       "      <td>9.717658</td>\n",
       "      <td>10.110915</td>\n",
       "      <td>7.609833</td>\n",
       "      <td>...</td>\n",
       "      <td>3.305795</td>\n",
       "      <td>6.091046</td>\n",
       "      <td>4.507657</td>\n",
       "      <td>7.767780</td>\n",
       "      <td>6.598748</td>\n",
       "      <td>5.564879</td>\n",
       "      <td>7.751093</td>\n",
       "      <td>7.101492</td>\n",
       "      <td>8.273221</td>\n",
       "      <td>5.316456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-101</th>\n",
       "      <td>7.073471</td>\n",
       "      <td>5.457774</td>\n",
       "      <td>7.012125</td>\n",
       "      <td>7.053250</td>\n",
       "      <td>4.402102</td>\n",
       "      <td>5.990521</td>\n",
       "      <td>6.686977</td>\n",
       "      <td>9.700598</td>\n",
       "      <td>10.393530</td>\n",
       "      <td>8.135075</td>\n",
       "      <td>...</td>\n",
       "      <td>3.223861</td>\n",
       "      <td>6.341113</td>\n",
       "      <td>4.207010</td>\n",
       "      <td>8.455561</td>\n",
       "      <td>7.629895</td>\n",
       "      <td>5.230453</td>\n",
       "      <td>8.856678</td>\n",
       "      <td>6.931344</td>\n",
       "      <td>8.200737</td>\n",
       "      <td>4.812053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-11</th>\n",
       "      <td>6.758017</td>\n",
       "      <td>4.729599</td>\n",
       "      <td>6.839936</td>\n",
       "      <td>7.046774</td>\n",
       "      <td>4.155380</td>\n",
       "      <td>6.883295</td>\n",
       "      <td>6.468586</td>\n",
       "      <td>9.358712</td>\n",
       "      <td>10.131947</td>\n",
       "      <td>7.556673</td>\n",
       "      <td>...</td>\n",
       "      <td>4.076035</td>\n",
       "      <td>5.730058</td>\n",
       "      <td>5.031616</td>\n",
       "      <td>7.723930</td>\n",
       "      <td>7.334398</td>\n",
       "      <td>5.295999</td>\n",
       "      <td>9.172799</td>\n",
       "      <td>6.864444</td>\n",
       "      <td>8.265620</td>\n",
       "      <td>4.839428</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18849 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0  ENSG00000000003  ENSG00000000419  ENSG00000000457  \\\n",
       "Sample-1           7.316251         4.274548         7.262058   \n",
       "Sample-10          6.808974         5.260135         7.115016   \n",
       "Sample-100         6.552535         4.804724         6.923650   \n",
       "Sample-101         7.073471         5.457774         7.012125   \n",
       "Sample-11          6.758017         4.729599         6.839936   \n",
       "\n",
       "Unnamed: 0  ENSG00000000460  ENSG00000000938  ENSG00000000971  \\\n",
       "Sample-1           6.663354         5.221161         6.813789   \n",
       "Sample-10          7.203029         4.165419         6.460773   \n",
       "Sample-100         6.245148         5.204919         6.162263   \n",
       "Sample-101         7.053250         4.402102         5.990521   \n",
       "Sample-11          7.046774         4.155380         6.883295   \n",
       "\n",
       "Unnamed: 0  ENSG00000001036  ENSG00000001084  ENSG00000001167  \\\n",
       "Sample-1           6.806842         9.458696        10.129412   \n",
       "Sample-10          6.812336         9.905395        10.617105   \n",
       "Sample-100         6.554665         9.717658        10.110915   \n",
       "Sample-101         6.686977         9.700598        10.393530   \n",
       "Sample-11          6.468586         9.358712        10.131947   \n",
       "\n",
       "Unnamed: 0  ENSG00000001460  ...  ENSG00000283597  ENSG00000283619  \\\n",
       "Sample-1           7.288785  ...         4.340793         5.641259   \n",
       "Sample-10          7.933976  ...         3.712015         5.428915   \n",
       "Sample-100         7.609833  ...         3.305795         6.091046   \n",
       "Sample-101         8.135075  ...         3.223861         6.341113   \n",
       "Sample-11          7.556673  ...         4.076035         5.730058   \n",
       "\n",
       "Unnamed: 0  ENSG00000283623  ENSG00000283633  ENSG00000283635  \\\n",
       "Sample-1           3.754220         7.674109         7.034334   \n",
       "Sample-10          4.434448         8.343289         7.447873   \n",
       "Sample-100         4.507657         7.767780         6.598748   \n",
       "Sample-101         4.207010         8.455561         7.629895   \n",
       "Sample-11          5.031616         7.723930         7.334398   \n",
       "\n",
       "Unnamed: 0  ENSG00000283646  ENSG00000283652  ENSG00000283662  \\\n",
       "Sample-1           5.624611         9.289292         6.550443   \n",
       "Sample-10          5.615707         8.571755         6.627103   \n",
       "Sample-100         5.564879         7.751093         7.101492   \n",
       "Sample-101         5.230453         8.856678         6.931344   \n",
       "Sample-11          5.295999         9.172799         6.864444   \n",
       "\n",
       "Unnamed: 0  ENSG00000283674  ENSG00000283683  \n",
       "Sample-1           7.916829         4.786196  \n",
       "Sample-10          7.709677         5.008819  \n",
       "Sample-100         8.273221         5.316456  \n",
       "Sample-101         8.200737         4.812053  \n",
       "Sample-11          8.265620         4.839428  \n",
       "\n",
       "[5 rows x 18849 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_merged=pd.read_csv('data\\\\metabolomics\\\\data.csv').set_index('Unnamed: 0').T\n",
    "d_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4a36a2cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>RNAIsolation.Group</th>\n",
       "      <th>Dissection.Group</th>\n",
       "      <th>Sample.ID</th>\n",
       "      <th>Case.Year</th>\n",
       "      <th>Case.Num</th>\n",
       "      <th>Region</th>\n",
       "      <th>Neuropath.Dx.1</th>\n",
       "      <th>Neuropath.Dx.2</th>\n",
       "      <th>Age</th>\n",
       "      <th>...</th>\n",
       "      <th>MEDIAN_5PRIME_TO_3PRIME_BIAS</th>\n",
       "      <th>TOTAL_CLUSTERS</th>\n",
       "      <th>ALIGNED_READS</th>\n",
       "      <th>AT_DROPOUT</th>\n",
       "      <th>GC_DROPOUT</th>\n",
       "      <th>Seq.PC1</th>\n",
       "      <th>Seq.PC2</th>\n",
       "      <th>Seq.PC3</th>\n",
       "      <th>Seq.PC4</th>\n",
       "      <th>Seq.PC5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sample-1</th>\n",
       "      <td>1</td>\n",
       "      <td>Group16</td>\n",
       "      <td>Group7</td>\n",
       "      <td>1</td>\n",
       "      <td>2000</td>\n",
       "      <td>8</td>\n",
       "      <td>FC</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>Unused</td>\n",
       "      <td>82</td>\n",
       "      <td>...</td>\n",
       "      <td>0.870195</td>\n",
       "      <td>42669514</td>\n",
       "      <td>65203198</td>\n",
       "      <td>20.803772</td>\n",
       "      <td>0.163871</td>\n",
       "      <td>-0.133179</td>\n",
       "      <td>-0.103026</td>\n",
       "      <td>0.412722</td>\n",
       "      <td>0.423658</td>\n",
       "      <td>-0.529204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-10</th>\n",
       "      <td>2</td>\n",
       "      <td>Group13</td>\n",
       "      <td>Group13</td>\n",
       "      <td>10</td>\n",
       "      <td>2002</td>\n",
       "      <td>4</td>\n",
       "      <td>FC</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>Unused</td>\n",
       "      <td>83</td>\n",
       "      <td>...</td>\n",
       "      <td>0.543860</td>\n",
       "      <td>26814831</td>\n",
       "      <td>46094008</td>\n",
       "      <td>3.724324</td>\n",
       "      <td>3.087543</td>\n",
       "      <td>0.027563</td>\n",
       "      <td>0.108425</td>\n",
       "      <td>-0.009570</td>\n",
       "      <td>0.085121</td>\n",
       "      <td>0.056185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-100</th>\n",
       "      <td>3</td>\n",
       "      <td>Group5</td>\n",
       "      <td>Group17</td>\n",
       "      <td>100</td>\n",
       "      <td>1999</td>\n",
       "      <td>26</td>\n",
       "      <td>FC</td>\n",
       "      <td>Normal (Mild Braak Changes)</td>\n",
       "      <td>Unused</td>\n",
       "      <td>79</td>\n",
       "      <td>...</td>\n",
       "      <td>0.567146</td>\n",
       "      <td>32645219</td>\n",
       "      <td>59133696</td>\n",
       "      <td>7.030072</td>\n",
       "      <td>1.990039</td>\n",
       "      <td>-0.064691</td>\n",
       "      <td>-0.033957</td>\n",
       "      <td>-0.093489</td>\n",
       "      <td>0.026464</td>\n",
       "      <td>-0.033087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-101</th>\n",
       "      <td>4</td>\n",
       "      <td>Group8</td>\n",
       "      <td>Group6</td>\n",
       "      <td>101</td>\n",
       "      <td>1999</td>\n",
       "      <td>37</td>\n",
       "      <td>FC</td>\n",
       "      <td>Normal (Mild Braak Changes)</td>\n",
       "      <td>Unused</td>\n",
       "      <td>74</td>\n",
       "      <td>...</td>\n",
       "      <td>0.737687</td>\n",
       "      <td>24149702</td>\n",
       "      <td>44736556</td>\n",
       "      <td>11.776616</td>\n",
       "      <td>1.562165</td>\n",
       "      <td>0.051882</td>\n",
       "      <td>-0.150986</td>\n",
       "      <td>-0.042881</td>\n",
       "      <td>-0.068372</td>\n",
       "      <td>-0.000339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample-11</th>\n",
       "      <td>13</td>\n",
       "      <td>Group2</td>\n",
       "      <td>Group17</td>\n",
       "      <td>11</td>\n",
       "      <td>2002</td>\n",
       "      <td>16</td>\n",
       "      <td>FC</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>Unused</td>\n",
       "      <td>90</td>\n",
       "      <td>...</td>\n",
       "      <td>0.455994</td>\n",
       "      <td>26803406</td>\n",
       "      <td>50174354</td>\n",
       "      <td>3.197977</td>\n",
       "      <td>3.365408</td>\n",
       "      <td>0.008923</td>\n",
       "      <td>0.018487</td>\n",
       "      <td>-0.126085</td>\n",
       "      <td>-0.020246</td>\n",
       "      <td>-0.061431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0 RNAIsolation.Group Dissection.Group  Sample.ID  \\\n",
       "SampleID                                                                \n",
       "Sample-1             1            Group16           Group7          1   \n",
       "Sample-10            2            Group13          Group13         10   \n",
       "Sample-100           3             Group5          Group17        100   \n",
       "Sample-101           4             Group8           Group6        101   \n",
       "Sample-11           13             Group2          Group17         11   \n",
       "\n",
       "            Case.Year  Case.Num Region               Neuropath.Dx.1  \\\n",
       "SampleID                                                              \n",
       "Sample-1         2000         8     FC          Alzheimer's disease   \n",
       "Sample-10        2002         4     FC          Alzheimer's disease   \n",
       "Sample-100       1999        26     FC  Normal (Mild Braak Changes)   \n",
       "Sample-101       1999        37     FC  Normal (Mild Braak Changes)   \n",
       "Sample-11        2002        16     FC          Alzheimer's disease   \n",
       "\n",
       "           Neuropath.Dx.2  Age  ... MEDIAN_5PRIME_TO_3PRIME_BIAS  \\\n",
       "SampleID                        ...                                \n",
       "Sample-1           Unused   82  ...                     0.870195   \n",
       "Sample-10          Unused   83  ...                     0.543860   \n",
       "Sample-100         Unused   79  ...                     0.567146   \n",
       "Sample-101         Unused   74  ...                     0.737687   \n",
       "Sample-11          Unused   90  ...                     0.455994   \n",
       "\n",
       "            TOTAL_CLUSTERS ALIGNED_READS AT_DROPOUT GC_DROPOUT   Seq.PC1  \\\n",
       "SampleID                                                                   \n",
       "Sample-1          42669514      65203198  20.803772   0.163871 -0.133179   \n",
       "Sample-10         26814831      46094008   3.724324   3.087543  0.027563   \n",
       "Sample-100        32645219      59133696   7.030072   1.990039 -0.064691   \n",
       "Sample-101        24149702      44736556  11.776616   1.562165  0.051882   \n",
       "Sample-11         26803406      50174354   3.197977   3.365408  0.008923   \n",
       "\n",
       "             Seq.PC2   Seq.PC3   Seq.PC4   Seq.PC5  \n",
       "SampleID                                            \n",
       "Sample-1   -0.103026  0.412722  0.423658 -0.529204  \n",
       "Sample-10   0.108425 -0.009570  0.085121  0.056185  \n",
       "Sample-100 -0.033957 -0.093489  0.026464 -0.033087  \n",
       "Sample-101 -0.150986 -0.042881 -0.068372 -0.000339  \n",
       "Sample-11   0.018487 -0.126085 -0.020246 -0.061431  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata=pd.read_csv('data\\\\metabolomics\\\\target.csv').set_index('SampleID')\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0974a6ac",
   "metadata": {},
   "source": [
    "### Setting paramerers for program"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f7e7ad48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#columns we consider as covariates\n",
    "covariates_columns=['Age', 'Sex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7cb73c71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18849"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#factores we want to study\n",
    "p=d_merged.columns\n",
    "len(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d82ef468",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_merged[covariates_columns]=metadata.loc[d_merged.index, covariates_columns].replace({'M':0, 'F':1}).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "066dd834",
   "metadata": {},
   "outputs": [],
   "source": [
    "#array with indexes for \"normal\" samples and \"disease\" samples\n",
    "N=metadata.index[metadata['Neuropath.Dx.1']==\"Alzheimer's disease\"]\n",
    "D=metadata.index[metadata['Neuropath.Dx.1']!=\"Alzheimer's disease\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "eaf3c134",
   "metadata": {},
   "outputs": [],
   "source": [
    "#check and fix batch effect on batch_column (the column indicates datasets)\n",
    "fix_batch_effect=False\n",
    "batch_column=None\n",
    "\n",
    "#which filter functions to use\n",
    "filters=[\n",
    "    'DSGA', \n",
    "    'KNearest', \n",
    "    'MDS', \n",
    "    'PCA', \n",
    "    'PCAflat'\n",
    "]\n",
    "\n",
    "#where to save results or look for previously culculated data\n",
    "folder_name='./images/AD/' \n",
    "\n",
    "#path to the K-nearest neighbors distance matrix if it have been calculated \n",
    "knn_dist_path=None\n",
    "\n",
    "#cut-off for normal samples to reduce number of samples to speed up calculations \n",
    "n_s1=None\n",
    "#cut-off for disease samples \n",
    "n_s2=None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa36756f",
   "metadata": {},
   "source": [
    "Parsing results if avalable to extract factors significatly differs in subgroupes of samples. This is needed for the persistance homology pipline to reduce dimention. We consider most wariable and significatly different features. You can specify other fetures in this list. If the empty list is given the same same function would be performed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5cd56955",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign=main.res_scrap(folder_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6845a8",
   "metadata": {},
   "source": [
    "### Perform calculations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd46e5a8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DSGA\n",
      "PH pipline\n",
      "KeplerMapper(verbose=1)\n",
      "Mapping on data shaped (90, 18851) using lens shaped (90, 2)\n",
      "\n",
      "Creating 225 hypercubes.\n",
      "\n",
      "Created 49 edges and 31 nodes in 0:00:00.394675.\n",
      "Something wrong 0\n",
      "KNearest\n",
      "PH pipline\n",
      "KeplerMapper(verbose=1)\n",
      "Mapping on data shaped (90, 18851) using lens shaped (90, 2)\n",
      "\n",
      "Creating 169 hypercubes.\n",
      "\n",
      "Created 242 edges and 82 nodes in 0:00:00.825070.\n"
     ]
    }
   ],
   "source": [
    "main.run(d_merged, D, N, list(p), list(covariates_columns), \n",
    "         folder_name, fix_batch_effect, batch_column,\n",
    "        filters,\n",
    "        knn_dist_path,\n",
    "        n_s1,\n",
    "        n_s2, \n",
    "        overlap_l=20, #lower edge to check for overlap in the Mapper\n",
    "        PH=True, #Either to perform the persistance homology pipline\n",
    "        sign=sign\n",
    "       )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
